cache/memCache.go

```diff
@@ -0,0 +1,116 @@
+package cache
+
+import (
+	"challenge/models"
+	"errors"
+	"runtime"
+	"time"
+)
+
+type memCacheHandler struct {
+	Block              map[string]models.CacheEntry // UserID: CacheEntry{}
+	DefaultCacheExpiry time.Duration
+	Released           bool
+}
+
+// createMemoryHandler initializes the memory structure for the users/assets.
+func createMemCacheHander() *memCacheHandler {
+	newMemCache := memCacheHandler{}
+	newMemCache.Block = make(map[string]models.CacheEntry)
+	// Setting default cache expiry at 20 minutes
+	newMemCache.DefaultCacheExpiry = 20 * time.Minute
+	return &newMemCache
+}
+
+// cacheEntryInvalidator checks the cache entry expiry and invalidates it when
+// it expires.
+func (m *memCacheHandler) cacheEntryInvalidator(entry *models.CacheEntry) {
+	for {
+		if m.Released {
+			return
+		}
+		if entry.ExpiresAt.Before(time.Now()) {
+			entry.Set = false
+			Cache.Evict(entry.Token)
+			return
+		} else {
+			time.Sleep(1 * time.Second)
+		}
+	}
+}
+
+// Store create a new cache entry mapping the response to the user token
+// This entry is replaced on favourites list changes.
+func (m *memCacheHandler) Store(token string, response []byte, invalidator bool, expiration ...time.Time) error {
+	if m.Released {
+		err := errors.New("cache resource is not allocated")
+		return err
+	}
+	newCacheEntry := models.CacheEntry{
+		Token:     token,
+		Response:  response,
+		Set:       true,
+		ExpiresAt: time.Now().Add(m.DefaultCacheExpiry),
+	}
+	if len(expiration) > 0 {
+		newCacheEntry.ExpiresAt = expiration[0]
+	}
+	// Start the invalidator counter (if requested)
+	if invalidator {
+		go m.cacheEntryInvalidator(&newCacheEntry)
```

Comment:
In this case they are not bound. They have two termination clauses (memory release, expiration) but manual setting of no expiration or a massive influx of new cache entry requests could impact performance. This can be mitigated by either having a single invalidator doing all the checks (various methods of implementation - brute force, channel messages, etc) which could lead to a delta between expiration time and eviction, or by having a set amount of goroutines performing this operation (tasks on buffered channels, etc).

Reply:
Thanks, mate! That was exactly what I was looking for 🚀